python/python/tests/test_dataset.py

```diff
@@ -614,29 +614,64 @@ def test_cleanup_old_versions(tmp_path):
     assert stats.old_versions == 1
 
 
-def test_do_not_cleanup_old_tagged_versions(tmp_path):
+def test_cleanup_error_when_tagged_old_versions(tmp_path):
     table = pa.Table.from_pydict({"a": range(100), "b": range(100)})
     base_dir = tmp_path / "test"
     lance.write_dataset(table, base_dir)
+    lance.write_dataset(table, base_dir, mode="overwrite")
     moment = datetime.now()
     lance.write_dataset(table, base_dir, mode="overwrite")
 
     dataset = lance.dataset(base_dir)
     dataset.create_tag("old-tag", 1)
-    dataset.create_tag("another-old-tag", 1)
-    dataset.create_tag("tag-latest", dataset.latest_version)
+    dataset.create_tag("another-old-tag", 2)
 
-    stats = dataset.cleanup_old_versions(older_than=(datetime.now() - moment))
-    assert stats.bytes_removed == 0
-    assert stats.old_versions == 0
+    with pytest.raises(OSError):
+        dataset.cleanup_old_versions(older_than=(datetime.now() - moment))
+    assert len(dataset.versions()) == 3
 
     dataset.delete_tag("old-tag")
+    with pytest.raises(OSError):
+        dataset.cleanup_old_versions(older_than=(datetime.now() - moment))
+    assert len(dataset.versions()) == 3
+
+    dataset.delete_tag("another-old-tag")
     stats = dataset.cleanup_old_versions(older_than=(datetime.now() - moment))
+    assert stats.bytes_removed > 0
+    assert stats.old_versions == 2
+    assert len(dataset.versions()) == 1
+
+
+def test_cleanup_around_tagged_old_versions(tmp_path):
+    table = pa.Table.from_pydict({"a": range(100), "b": range(100)})
+    base_dir = tmp_path / "test"
+    lance.write_dataset(table, base_dir)
+    lance.write_dataset(table, base_dir, mode="overwrite")
+    moment = datetime.now()
+    lance.write_dataset(table, base_dir, mode="overwrite")
+
+    dataset = lance.dataset(base_dir)
+    dataset.create_tag("old-tag", 1)
```

Comment:
One API thought: it might be nice to consolidate the methods into a namespace. So

* `dataset.create_tag("old-tag", 1)` -> `dataset.tags.create("old-tag", 1)`
* `dataset.delete_tag("old-tag")` -> `dataset.tags.delete("old-tag")`
* `dataset.tags()` -> `datataset.tags.list()`

See the `Dataset.optimize` API for inspiration here.

Reply:
Did you want this implementing on the Rust side as well? Python is straightforward via the use of `@property` but this turned out to be harder than I expected and I haven't found a good way to do it in Rust yet. So far, I've tried:

- `dataset.tags` via a nested `Tags { object_store, commit_handler, base }` struct that is created at the same time as the `Dataset` struct. This works fine, but risks getting decoupled from the main `Dataset` instance if a variable like `base_path` changes. This option is currently implemented in this PR.
- `dataset.tags` via a nested `Tags { dataset: Arc<Dataset> }` struct that is created at the same time as the `Dataset` struct. This requires temporarily instantiating an empty reference (to create the main `Dataset` struct), before updating the reference to point back at `Dataset`. This requires `unsafe` code that seems overly complicated for what we are trying to achieve.
- A `dataset.tags(&self)` method which uses `self` to create a new `Tags { Arc<Dataset> }` struct. This messes with the API a little bit (e.g., `dataset.tags().list()` instead of `dataset.tags.list()` and requires a new struct to be created every time a tag operation is invoked.

Should we even be trying to namespace the Rust API? If so, how do you think this should be achieved?